File: array_of_zeroes.rb
# http://programmingpraxis.com/2014/11/21/an-array-of-zeroes/

def move_zeroes(array)
  right = (array.length - 1)
  left = 0
  until right == left
    case
    when array[left] == 0 && array[right] != 0
      array[left], array[right] = array[right], array[left]
      right -= 1
      left += 1
    when array[left] == 0 && array[right] == 0
      right -= 1
    when array[left] != 0 && array[right] != 0
      left += 1
    when array[left] != 0 && array[right] == 0
      left += 1
      right -= 1
    end
  end
  array
end

test_array = [1,0,2,0,0,3,4]
p move_zeroes(test_array)

File: array_of_zeroes.swift
// http://programmingpraxis.com/2014/11/21/an-array-of-zeroes/

let testArray = [1,0,2,0,0,3,4]

func moveZeroes(testArray:[Int]) -> [Int] {
  var mutableArray = testArray
  var low = 0
  var high = (mutableArray.count - 1)
  while low < high {
    if mutableArray[low] == 0 && mutableArray[high] != 0 {
      swap(&mutableArray[low], &mutableArray[high])
      ++low
      --high
    } else if mutableArray[low] == 0 && mutableArray[high] == 0 {
      --high
    } else if mutableArray[low] != 0 {
      ++low
    }
  }
  return mutableArray
}

println(moveZeroes(testArray))
